pkg/migration/migrations/grant_option_migration.go

```diff
@@ -44,6 +44,23 @@ func grantOptionMigration(
 	addGrantOptionFunc := func(ids []descpb.ID, descs []descpb.Descriptor, timestamps []hlc.Timestamp) error {
 		var modifiedDescs []catalog.MutableDescriptor
 		for i, id := range ids {
+			// Temporarily change the version in order to force the migration to run
```

Comment:
> Doesn't this make it so the migration can run twice - ie we don't have a base case for the descriptor state to reach?

Sort of -- but the migration itself only runs when the clusterversion reaches `clusterversion.ValidateGrantOption` and that can only happen once.

> Also are we sure the grant option bits are wiped out?
Yes, this was causing line 178 to fail here in validate_grant_option.go  ( https://github.com/cockroachdb/cockroach/pull/72982/commits/aafdb46c32751e956cf4180e3bef1b8cdffd7fe9#diff-edd6b676708815ac3b09d3f07bb9c8c4bffe7441dacc9452c6f7453c2ba194efR178 )

Reply:
here's the reason it gets wiped out:

- suppose you have a mixed-version cluster
- run `CREATE TABLE t` on a node running the new binary. this creates a privilege descriptor with a version of `GrantOptionVersion`.
- run `GRANT GRANT, SELECT ON t TO testuser` on a node running the new binary. `testuser` can now grant SELECT to others, and the descriptor has its grant options bits set.
- run `GRANT DELETE ON t TO testuser` on a node running the old binary. now the descriptor has the GRANT, SELECT, and DELETE privileges, but the grant option bits are not set since the old binary cannot serialize a descriptor with that info.
- finalize the cluster upgrade. without my change here, the grant option bits are not added, since the privilege descriptor already is at `GrantOptionVersion`. now `testuser` can NOT grant SELECT to others, even though they could right before the upgrade was finalized.

Reply:
Ah I see, it's not wiping out the grant bit but the migration isn't happening as expected.

This migration is kind of thorny, I'm still a bit iffy on downgrading the version temporarily, it just seems scary since we generally always expect the version upgrade to increase.

What if in this step
```
run GRANT GRANT, SELECT ON t TO testuser on a node running the new binary. testuser can now grant SELECT to others, and the descriptor has its grant options bits set.
```
We just also lazily upgrade the descriptor to have GRANT bits, and continue keeping GRANT privilege as well as we still need it until we finalize the migration, then in the next version we'll have a migration if necessary to then fully remove GRANT option (we can't do it til then)

Reply:
> it just seems scary since we generally always expect the version upgrade to increase.

well, the version is only modified in memory here; we never write that back to disk.

> We just also lazily upgrade the descriptor to have GRANT bits, and continue keeping GRANT privilege as well as we still need it until we finalize the migration

This is how it works right now.

>  it's not wiping out the grant bit but the migration isn't happening as expected.

It is wiping out the grant option bits

Reply:
I view the temporary, in-memory change to the version as safe. It's just a way to avoid having to do another migration in the 22.1->22.2 upgrade. with the migration like this, in during the 22.2 we should be able to just stop referencing the GRANT privilege in the code without needing to do anything else. (we'd still have to keep the GRANT enum value, so that it never accidentally gets reused in the future.)

Reply:
Sorry, I misunderstood. I see now, we moved some of the logic back into `RunPostDeserializedChanges`.

I'm still somewhat confused on how the grant bits get wiped, I thought since on the old node we wouldn't be updating the grant bits in the proto, they'd stay the same. Gonna check out the branch to double check my understanding.

My concern with downgrading the version is that it's kind of strange that we're doing the migration on privilege descriptors that shouldn't need the migration.

Ideally for me, new descriptors should not need the migration at all if they're created on that version.
Can we version gate the descriptor version as well? Only set the descriptor version to `GrantOptionVersion` iff the cluster verison is `ValidateGrantOption`?

Reply:
> I'm still somewhat confused on how the grant bits get wiped, I thought since on the old node we wouldn't be updating the grant bits in the proto, they'd stay the same.

it's because the old node's proto definition doesn't even have a field for the grant option bits. so when it reads the bytes and turns it into a privilege descriptor struct, it doesn't have that field set at all. and when it turns it back into bytes, it wouldn't even know how to serialize that field anyway.

> new descriptors should not need the migration at all if they're created on that version.
Can we version gate the descriptor version as well? Only set the descriptor version to GrantOptionVersion iff the cluster verison is ValidateGrantOption?

yes this would work too. i didn't do it though since it would require a lot of plumbing, which we'd then want to undo in 22.2.

> My concern with downgrading the version is that it's kind of strange that we're doing the migration on privilege descriptors that shouldn't need the migration.

Well the thing is, they do need the migration. If we go with your suggestion of setting GrantOptionVersion iff the cluster verison is ValidateGrantOption, then the migration would work the exact same way, in practice -- it would run on all privilege descriptors.

if it helps, i can do one more check for this "temporary version change" bit i added -- it only needs to run IFF there are no grant option bits currently set.

Reply:
Yeah looking into it, the plumbing is pretty tedious. Let me think about it for a little more to see if we can avoid it, otherwise it's fine.